src/python_testing/mdns_discovery/mdns_discovery.py

```diff
@@ -0,0 +1,318 @@
+
+        Args:
+            discovery_duration (float): The duration in seconds for which the discovery
+                                        process should wait to allow for service announcements.
+                                        If not provided, defaults to 3 seconds.
+
+        Note:
+            The discovery duration may need adjustment based on network conditions and the
+            expected response time for service announcements.
+
+        Returns:
+            None: This method does not return any value. The results of the service discovery
+                  are handled by the callback function.
+        """
+        self._discovered_services = {mdns_type.name: [] for mdns_type in MdnsType}
+        AsyncServiceBrowser(zeroconf=self._zc,
+                            type_=self._service_types,
+                            handlers=[self.on_service_state_change]
+                            )
+        # Wait for discovery
+        await asyncio.sleep(discovery_duration)
+        self._discovery_performed = True
+
+    def on_service_state_change(
+        self,
+        zeroconf: Zeroconf,
+        service_type: str,
+        name: str,
+        state_change: ServiceStateChange
+    ) -> None:
+        """
+        Callback method triggered on mDNS service state change.
+
+        This method is called by the Zeroconf library when there is a change in the state of an mDNS service.
+        It handles the addition of new services by initiating a query for their detailed information.
+
+        Args:
+            zeroconf (Zeroconf): The Zeroconf instance managing the network operations.
+            service_type (str): The type of the mDNS service that changed state.
+            name (str): The name of the mDNS service.
+            state_change (ServiceStateChange): The type of state change that occurred.
+
+        Returns:
+            None: This method does not return any value.
+        """
+        if state_change.value == ServiceStateChange.Added.value:
+            asyncio.ensure_future(self.query_service_info(
+                zeroconf,
+                service_type,
+                name)
+            )
+
+    async def query_service_info(self, zeroconf: Zeroconf, service_type: str, name: str) -> None:
+        """
+        This method requests detailed information about an mDNS service, such as its address,
+        port, and other service metadata. The information is requested asynchronously.
+
+        Args:
+            zeroconf (Zeroconf): The Zeroconf instance managing the network operations.
+            service_type (str): The type of the mDNS service to query.
+            name (str): The name of the mDNS service.
+
+        Returns:
+            None: This method does not return any value. The results of the query are processed internally.
+        """
+        # Get service info
+        service_info: AsyncServiceInfo = AsyncServiceInfo(service_type, name)
```

Comment:
I don't think typing this variable is needed as it obviously calls the constructor